twoliter/embedded/img2img

```diff
@@ -149,25 +162,29 @@ install_root_json "${ROOT_MOUNT}"
 ###############################################################################
 # Section 4: update root partition and root verity
 
-# shellcheck disable=SC2312  # mapfile is validated elsewhere
-mapfile -t new_root_artifacts <<<"$(find "${ROOT_MOUNT}" -type f)"
-
-# The reason we check index 0 rather than the mapfile length is if `find` fails
-# to find an artifact the heredoc to mapfile will assign empty output to 0.
-if [[ -z "${new_root_artifacts[0]}" ]]; then
-  echo "no new root artifacts found" >&2
-  exit 1
+if [[ "${EROFS_ROOT_PARTITION}" == "yes" ]]; then
+  mkfs_root_erofs "${ROOT_MOUNT}" "${ROOT_IMAGE}" "${SELINUX_FILE_CONTEXTS}"
 else
-  # Write files from the root mount to the root image.
-  ROOT_DEBUGFS_STDERR="${WORKDIR}/root.err"
-  for artifact in "${new_root_artifacts[@]}"; do
-    cat <<EOF | debugfs -w -f - "${ROOT_IMAGE}" 2>>"${ROOT_DEBUGFS_STDERR}"
+  # shellcheck disable=SC2312  # mapfile is validated elsewhere
+  mapfile -t new_root_artifacts <<<"$(find "${ROOT_MOUNT}" -type f)"
+
+  # The reason we check index 0 rather than the mapfile length is if `find` fails
+  # to find an artifact the heredoc to mapfile will assign empty output to 0.
+  if [[ -z "${new_root_artifacts[0]}" ]]; then
+    echo "no new root artifacts found" >&2
+    exit 1
+  else
+    # Write files from the root mount to the root image.
+    ROOT_DEBUGFS_STDERR="${WORKDIR}/root.err"
+    for artifact in "${new_root_artifacts[@]}"; do
+      cat <<EOF | debugfs -w -f - "${ROOT_IMAGE}" 2>>"${ROOT_DEBUGFS_STDERR}"
 rm ${artifact#"${ROOT_MOUNT}"}
 write ${artifact} ${artifact#"${ROOT_MOUNT}"}
 ea_set ${artifact#"${ROOT_MOUNT}"} security.selinux system_u:object_r:os_t:s0
 EOF
-  done
-  check_debugfs_errors "${ROOT_DEBUGFS_STDERR}"
+    done
+    check_debugfs_errors "${ROOT_DEBUGFS_STDERR}"
+  fi
```

Comment:
Since we're just completely remaking the root filesystem if it's erofs, should we follow the same workflow for ext4?

Reply:
For erofs, extracting the filesystem is a supported feature of `fsck.erofs`. For ext4, I'm more confident that `debugfs` will do what we want, since I'm not aware of a way to do full filesystem extraction using tools from e2fsprogs.

Reply:
The [erofs paper](https://www.usenix.org/system/files/atc19-gao.pdf) mentions support for "image patching" in 4.3, but I didn't find a way to do that with `erofs-utils`.